services/business/namespaces.go

```diff
@@ -0,0 +1,34 @@
+package business
+
+import (
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/services/models"
+)
+
+// Namespace deals with fetching k8s namespaces / OpenShift projects and convert to kiali model
+type NamespaceService struct {
+	k8s kubernetes.IstioClientInterface
+}
+
+// Returns a list of the given namespaces / projects
+func (in *NamespaceService) GetNamespaces() ([]models.Namespace, error) {
+	// First try and see if we are running on OpenShift of Kubernetes with the kube-project component.
+	// This will allow us to fetch the namespaces that only our serviceaccount has permission to access.
+	//
+	// If we encounter an error here, we are assuming its because the project.openshift.io component is unknown
+	// and it means we are running in k8s without kube-projet. In that case, get the namespace list from Kubernetes
+	// (which requires a higher permission level)
+	projects, err := in.k8s.GetProjects()
```

Comment:
This logic is too aggressive to be queried always IMO.
I mean, perhaps we can perform a 1-time GetProjects() when Kiali startup to check if it is under OpenShift or Kubernetes.
Falling into two queries per k8s scenarios sounds not necessary as it is something is not going to change between queries.

Reply:
+1, that check should be done one time most likely.